### ISSUE TITLE
Separate exn types

### ISSUE THREAD

**Issue by okeuday** [githwxi/ATS-Xanadu#27]
It would be helpful to have two separate types for exceptions, with one type for recoverable exceptions and the other type for unrecoverable exceptions (i.e., fatal exceptions).  The cleanest approach of this idea appears to be the Java class hierarchy with https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html for recoverable exceptions and https://docs.oracle.com/javase/8/docs/api/index.html?java/lang/Error.html for unrecoverable exceptions.  Having that clear delineation helps to communicate the intent in the source code for potential failures.

My impression is that at least the ATS2 exceptions `AssertExn` and `IllegalArgExn` are unrecoverable because they represent a broken constraint in the source code.

**Comment by Necryotiks** [comment]
Recoverable exceptions are usually handled by Option or Result types.

**Comment by okeuday** [comment]
@Necryotiks Yes, for referential transparency, that is preferred in any programming language.  This issue was focused on "if exceptions exist, how should they be" because of experiences with ATS2 exceptions.  The  `exit` C function is seen as an `exn` side-effect and the `atexit` C function registers a function pointer that is similar to a catch of an unrecoverable exception (without `exit` existing as an exception), so the unrecoverable exceptions concept is common/old/useful even if it doesn't use exception types.
